runtime/src/mem/jemalloc/mem-jemalloc.c

```diff
@@ -30,44 +30,198 @@
-  void* start;
-  size_t size;
 
-  chpl_comm_desired_shared_heap(&start, &size);
+//TODO move the chunk replacement code into a separate file
 
+// TODO put these into a struct?
+void*  heap_base = NULL;
+size_t heap_size = 0;
+size_t cur_heap_offset = 0;
+
+// *** Chunk hook replacements *** //
+// See http://www.canonware.com/download/jemalloc/jemalloc-latest/doc/jemalloc.html#arena.i.chunk_hooks
+
+// Our chunk replacement hook for allocations. Grab memory out of the shared
+// heap and give it to jemalloc.
+static void* chunk_alloc(void *chunk, size_t size, size_t alignment, bool *zero, bool *commit, unsigned arena_ind) {
+
+  void* p = NULL;
+  size_t p_offset;
+
+  // TODO cleanup some of the pointer arithmetic. The current code was
+  // originally copied out of the tcmalloc shim, and could use a little cleanup
+
+  // compute our current pointer into the shared heap based on the current
+  // offset aligned to "alignment"
+  p = (void*)((((uintptr_t) heap_base + cur_heap_offset + alignment - 1) / alignment) * alignment);
+
+  // """
+  // If chunk is not NULL, the returned pointer must be chunk on success or
+  // NULL on error
+  // """
+  if (chunk && chunk != p) {
+    return NULL;
+  }
+
+  p_offset = (uintptr_t)p - (uintptr_t)(heap_base);
+
+  size = ((size + alignment - 1) / alignment) * alignment;
+
+  // If we're out of space on the heap, return NULL
+  if (size > heap_size - p_offset) {
+    return NULL;
+  }
+
+  // Update the current pointer, now that we've past any early returns.
+  cur_heap_offset = p_offset + size;
+
+  // Zero if we were told to
+  if (*zero) {
+     memset(p, 0, size);
+  }
+
+  // Commit is not relevant for linux/darwin, but jemalloc wants us to set it
+  *commit = true;
+
+  return p;
+}
+
+//
+// Returning true indicates an opt-out of these hooks. For dalloc, this means
+// that we never get memory back from jemalloc and we just let it re-use it in
+// the future.
+//
+static bool null_dalloc(void *chunk, size_t size, bool committed, unsigned arena_ind) {
+  return true;
+}
+static bool null_commit(void *chunk, size_t size, size_t offset, size_t length, unsigned arena_ind) {
+  return true;
+}
+static bool null_decommit(void *chunk, size_t size, size_t offset, size_t length, unsigned arena_ind) {
+  return true;
+}
+static bool null_purge(void *chunk, size_t size, size_t offset, size_t length, unsigned arena_ind) {
+  return true;
+}
+static bool null_split(void *chunk, size_t size, size_t size_a, size_t size_b, bool committed, unsigned arena_ind) {
+  return true;
+}
+static bool null_merge(void *chunk_a, size_t size_a, void *chunk_b, size_t size_b, bool committed, unsigned arena_ind) {
+  return true;
+}
+// *** End chunk hook replacements *** //
+
+
+// TODO check for errors on all je_mallctl calls
+
+// get the number of arenas could use
+static unsigned get_num_arenas(void) {
+  unsigned narenas;
+  size_t sz;
+  sz = sizeof(narenas);
+  je_mallctl("opt.narenas", &narenas, &sz, NULL, 0);
```

Comment:
Just to follow up on this, I don't think we're going to be able to do this with the JE_MALLOC_CONF  env var. I think that really has to be set before the program is executed because jemalloc seems to do at least some initial setup even before a memory allocation routine is called. 

e.g. the following does not work:


```c
#include <stdio.h>
#include <stdlib.h>
#include <jemalloc/jemalloc.h>


// Setting this will change the number of arenas, but it has to be a compile
// time const
//const char * je_malloc_conf = "narenas:1";

static unsigned get_num_arenas(void) {
  unsigned narenas;
  size_t sz;
  sz = sizeof(narenas);
  je_mallctl("opt.narenas", &narenas, &sz, NULL, 0);
  return narenas;
}


// opt.narenas is a read only var, so this doesn't work
static void set_num_arenas(unsigned narenas) {
  je_mallctl("opt.narenas", NULL, NULL, &narenas, sizeof(narenas));
}

int main() {
  // does not work, jemalloc seems to have done some initial setup already
  setenv("JE_MALLOC_CONF", "narenas:4", 1);

  printf("Using %u arenas\n", get_num_arenas());
  void * p = je_malloc(8);
}

```